// @keystone-next/keystone/___internal-do-not-use-will-break-in-patch/next-graphql/dist/keystone-next-keystone-___internal-do-not-use-will-break-in-patch-next-graphql.esm.js
import { i as initConfig, c as createSystem } from '../../../dist/initConfig-fe2fa91d.esm.js';
import { c as createApolloServerMicro } from '../../../dist/createApolloServer-bb8f01e4.esm.js';
import '@babel/runtime/helpers/objectSpread2';
import '../../../dist/core-9f44263b.esm.js';
import '../../../dist/sqlite-ae6a913c.esm.js';
import '../../../dist/graphql-ts-schema-74a8c4fd.esm.js';
import '@graphql-ts/schema';
import 'graphql-type-json';
import 'graphql-upload/public/GraphQLUpload.js';
import 'graphql';
import 'decimal.js';
import '@graphql-ts/schema/api-without-context';
import '@graphql-ts/extend';
import '@graphql-ts/schema/api-with-context';
import '@graphql-tools/schema';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../../dist/graphql-errors-7d879e1a.esm.js';
import 'apollo-server-errors';
import 'path';
import 'util';
import 'fs-extra';
import 'fast-glob';
import 'resolve';
import '@nodelib/fs.walk';
import '@emotion/hash';
import '../../../dist/admin-meta-graphql-433469b3.esm.js';
import '@apollo/client';
import '@babel/runtime/helpers/classPrivateFieldInitSpec';
import '@babel/runtime/helpers/classPrivateFieldGet';
import '@babel/runtime/helpers/classPrivateFieldSet';
import 'p-limit';
import '../../../dist/types-for-lists-4d9c28b5.esm.js';
import 'graphql/execution/values';
import 'pluralize';
import '@babel/runtime/helpers/defineProperty';
import 'uuid';
import 'image-type';
import 'image-size';
import '../../../fields/types/image/utils/dist/keystone-next-keystone-fields-types-image-utils.esm.js';
import 'crypto';
import 'stream';
import 'filenamify';
import '@sindresorhus/slugify';
import '../../../fields/types/file/utils/dist/keystone-next-keystone-fields-types-file-utils.esm.js';
import 'node-fetch';
import 'form-data';
import 'cuid';
import '../../../dist/package-path-edbecc32.esm.js';
import 'apollo-server-micro';
import 'apollo-server-express';
import 'apollo-server-core';
import '../../../session/dist/keystone-next-keystone-session.esm.js';
import 'cookie';
import '@hapi/iron';
import 'uid-safe';

function nextGraphQLAPIRoute(keystoneConfig, prismaClient) {
  var _keystoneConfig$graph;

  const initializedKeystoneConfig = initConfig(keystoneConfig);
  const {
    graphQLSchema,
    getKeystone
  } = createSystem(initializedKeystoneConfig);
  const keystone = getKeystone(prismaClient);
  keystone.connect();
  const apolloServer = createApolloServerMicro({
    graphQLSchema,
    createContext: keystone.createContext,
    sessionStrategy: initializedKeystoneConfig.session,
    graphqlConfig: initializedKeystoneConfig.graphql,
    connectionPromise: keystone.connect()
  });
  return apolloServer.createHandler({
    path: ((_keystoneConfig$graph = keystoneConfig.graphql) === null || _keystoneConfig$graph === void 0 ? void 0 : _keystoneConfig$graph.path) || '/api/graphql'
  });
}

export { nextGraphQLAPIRoute };
